protocols/threshold-bls/src/state_machine/keygen.rs

```diff
@@ -0,0 +1,216 @@
+#[derive(Clone, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct MsgReliabilityCheck<D: Digest>(pub digest::Output<D>);
+
+#[derive(udigest::Digestable)]
+#[udigest(tag = "gennaro.dkg.keygen.tag")]
+enum Tag<'a> {
+    /// Tag that includes the prover index
+    Indexed {
+        party_index: u16,
+        #[udigest(as_bytes)]
+        sid: &'a [u8],
+    },
+    /// Tag w/o party index
+    Unindexed {
+        #[udigest(as_bytes)]
+        sid: &'a [u8],
+    },
+}
+
+pub async fn run_threshold_keygen<G, R, M, L, D>(
+    mut tracer: Option<&mut dyn Tracer>,
+    i: u16,
+    t: u16,
+    n: u16,
+    reliable_broadcast_enforced: bool,
+    execution_id: ExecutionId<'_>,
+    rng: &mut R,
+    party: M,
+) -> Result<(), Error>
+where
+    G: Group + GroupEncoding + Default,
+    D: Digest + Clone + 'static,
+    R: RngCore + CryptoRng,
+    M: Mpc<ProtocolMessage = Msg<G, D>>,
+{
+    tracer.protocol_begins();
+    let parameters = Parameters::new(
+        NonZeroUsize::new(t.into()).unwrap(),
+        NonZeroUsize::new(n.into()).unwrap(),
+    );
+    let mut participant =
+        SecretParticipant::<G>::new(NonZeroUsize::new(1).unwrap(), parameters).unwrap();
+
+    tracer.stage("Setup networking");
+    let MpcParty { delivery, .. } = party.into_party();
+    let (incomings, mut outgoings) = delivery.split();
+
+    let mut rounds = RoundsRouter::<Msg<G, D>>::builder();
+    let round1_broad = rounds.add_round(RoundInput::<MsgRound1Broadcast<G>>::broadcast(i, n));
+    let round1_p2p = rounds.add_round(RoundInput::<MsgRound1P2P>::p2p(i, n));
+    let round2 = rounds.add_round(RoundInput::<MsgRound2>::broadcast(i, n));
+    let round3 = rounds.add_round(RoundInput::<MsgRound3<G>>::p2p(i, n));
+    let round4 = rounds.add_round(RoundInput::<MsgRound4<G>>::broadcast(i, n));
+    let mut rounds = rounds.listen(incomings);
+
+    // Round 1
+    tracer.round_begins();
+    tracer.stage("Compute execution id");
+    let sid = execution_id.as_bytes();
+    let tag = |j| {
+        udigest::Tag::<D>::new_structured(Tag::Indexed {
+            party_index: j,
+            sid,
+        })
+    };
+    let tag_i = tag(i);
+
+    tracer.send_msg();
+    let round1_msgs: (Round1BroadcastData<G>, BTreeMap<usize, Round1P2PData>) =
+        match participant.round1() {
+            Ok(msgs) => msgs,
+            Err(e) => {
+                return Err(Error::RoundError(
+                    1,
+                    "Failed to create round 1 broadcast message".to_string(),
+                ))
+            }
+        };
+    outgoings
+        .send(Outgoing::broadcast(Msg::Round1Broadcast(
+            MsgRound1Broadcast { msg: round1_msgs.0 },
+        )))
+        .await
+        .map_err(|e| {
+            Error::IoError(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                e.to_string(),
+            ))
+        })?;
+    tracer.msg_sent();
+
+    // Round 2
+    tracer.round_begins();
```

Comment:
This code looks like blocking code. We will need to wrap all blocking code inside a `tokio::task::spawn_blocking` closure to ensure high performance. Basically every `tracer.X` function should be wrapped.

Reply:
So this is the type of logic I found inside the cggmp's implementation of keygen. Can you verify all that's needed here @tbraun96. At some point I want you to take this over. Mostly trying to establish here some additional flows for non-Zengo MPC libraries, since it seems like we may want to easily hook anything up to round-based.

- https://github.com/dfns/cggmp21/blob/m/cggmp21/src/keygen/threshold.rs#L120